Add postfix snippet extra

zdcthomas opened the pull request:
In a conversation over on issues #457, @L3MON4D3 and I discussed various ways
to best support postfix snippets. @L3MON4D3 came up with the initial callback
that supports this pr, thanks!!

This tries to handle all options intelligently, but if I missed something, please let me know!

questions:
- [x] right now, when a line doesn't match, I'm handing that by passing an empty string as `snippet.pre_dot` but I also toyed around with the idea of not letting the snippet expand. What do you think?

leiserfg commented:
Not sure if `pre_dot` is the best name for the function, sims like it's tied to the fact that the prefix is supposed to start with a `.` but that is not necessarily true.

L3MON4D3 commented:
Nice, that looks good so far :+1:

I agree with @leiserfg, maybe `postfix_text` or `postfix_content`?

Another idea: it would be nice to put this value into `env`, that way we 
* don't "clutter" the snippet (and `snip.env.SOME_VAR` seems nicer than `snip.some_var`) and
* can actually use it in lambdas, which would be pretty cool (`l("("..l.POSTIFIX_TEXT..")", {})` is more readable than the function-node-version)

Actually implementing that would require some way to set/extend the `env` from the `pre_expand`-callback, I think it should be alright to just return the extensions.
I'll look into how to properly do that, sounds fun :D

L3MON4D3 commented:
@zdcthomas you can play around with #464, I'll let it sit for a bit to see if there are better ways to modify the `env`, but this works

zdcthomas commented:
That looks great! I'll merge that in an try to incorporate it.

Also, it looked like the last workflow failed and it seemed like it was because I modified the luasnip.txt directly. I moved those changes to the DOC.md so hopefully that can run successfully now, but let me know if I need to change anything else.

I'm also going to take a crack at writing some tests!

L3MON4D3 commented:
CI was acting up lately, stupid workflows, but I think they should be okay now.

Tests :+1::+1:

L3MON4D3 commented on lua/luasnip/extras/postfix.lua:
```diff
@@ -0,0 +1,75 @@
+local events = require("luasnip.util.events")
+local matches = {
+	default = [[[%w%.%_%-%"%']+$]],
+	line = "^.+$",
```
Does this have any use currently?

zdcthomas replied:
no, but I was thinking this section could be a set of commonly used matches, for people to plug in to the postfix snippets. We could totally get rid of it though

L3MON4D3 replied:
Ahh, I didn't see that it was exported.
Let's keep it, it should be useful, would you mention that in the doc as well?

L3MON4D3 commented on lua/luasnip/extras/postfix.lua:
```diff
@@ -0,0 +1,75 @@
+	end
+
+	context = vim.tbl_deep_extend("keep", context, { wordTrig = false })
+	opts = vim.tbl_deep_extend(
+		"force",
+		opts,
+		generate_opts(match_pattern, user_callback)
+	)
```
This could be replaced with a single set on `callbacks.-1.pre_expand`, right? Would be a tiny bit clearer IMO

zdcthomas replied:
Sounds good to me!

zdcthomas replied:
Is there a nicer way to do this than a couple of nested nil checks?

L3MON4D3 replied:
Not really, but those could be moved into a separate function, `luasnip.util.util.tbl_set`,with an interface similar to `vim.tbl_get`, to keep the code readable

zdcthomas commented:
I'm having a small issue with the tests. My test looks like
```lua
local helpers = require("test.functional.helpers")(after_each)
local exec_lua, feed, exec = helpers.exec_lua, helpers.feed, helpers.exec
local ls_helpers = require("helpers")
local Screen = require("test.functional.ui.screen")

describe("postfix snippets", function()
	local screen

	before_each(function()
		helpers.clear()
		ls_helpers.session_setup_luasnip()

		screen = Screen.new(50, 3)
		screen:attach()
		screen:set_default_attr_ids({
			[0] = { bold = true, foreground = Screen.colors.Blue },
			[1] = { bold = true, foreground = Screen.colors.Brown },
			[2] = { bold = true },
			[3] = { background = Screen.colors.LightGray },
		})
	end)

	after_each(function()
		screen:detach()
	end)

	it(
		"creates a postfix snippet which changes the previous text once expanded",
		function()
			local postfix_snip = [[
	    pf(".parens", {
	      f(function(_, parent)
	        return "(" .. parent.postfix_match .. ")"
        end, {})
      })
	  ]]

			feed("bar.parens")
			exec_lua("ls.snip_expand(" .. postfix_snip .. ")")
			screen:expect({
				grid = [[
			(bar)^                                        |
			{0:~                                                 }|
			{2:-- INSERT --}                                      |]],
			})
		end
	)
end)
```
but the test is failing with this error message
```
FAILED   ../../tests/integration/postfix_snippet_spec.lua @ 27: postfix snippets creates a postfix snippet which changes the previous text once expanded
../../tests/integration/postfix_snippet_spec.lua:40: Row 1 did not match.
Expected:
  |*(bar)^                                        |
  |{0:~                                                 }|
  |{2:-- INSERT --}                                      |
Actual:
  |*(r.parens)^                                        |
  |{0:~                                                 }|
  |{2:-- INSERT --}                                      |
```

zdcthomas commented:
Looking more closely, it looks like none of the tests feed the trigger before expanding. If I match that behavior, I get the expected test result, but I'd like to understand why. I'll keep digging to understand more

zdcthomas commented:
Alright, I think that's pretty good for testing at the moment. Anything else y'all can see that needs changing? (Thanks for the patience with the first time contributor questions!)

L3MON4D3 commented:
> Looking more closely, it looks like none of the tests feed the trigger before expanding. If I match that behavior, I get the expected test result, but I'd like to understand why. I'll keep digging to understand more

Not sure if you've figured it out, in case not, the snippets don't need their trigger if they are `snip_expand`ed (of course there are exceptions, if they need `expand_params` for example)

L3MON4D3 commented:
Ah, as for your question(which I've managed to completely miss so far, sorry😅): I think it's alright to pass an empty string, it shouldn't occur during regular usage, and if it happens through an error I'd rather that something unexpected happens, than nothing.

L3MON4D3 commented on DOC.md:
```diff
@@ -378,6 +379,81 @@ s("trig", {
+
+      postfix(".br", {
+              f(function(_, parent)
+                    return "[" .. parent.postfix_match .. "]"
```
Would you update this to env (ofc) and use `parent.snippet` instead of `parent`? (in this case it's the same table, but there are snippets where they are not, and I'd rather the examples still work then).
Also, could you add one example with lambda? (is it obvious I like them yet? :P)

zdcthomas replied:
Oh crap I missed one! 
And absolutely about the lambda lol :)

zdcthomas replied:
Alright, Those examples are in!

L3MON4D3 replied:
Nice, that looks good :+1:

L3MON4D3 commented:
Ah, one last thing, `postfix_match` should be `POSTFIX_MATCH` I think, just to be more in line with the other variables in `env` (`TM_SELECTED_TEXT`, `UUID` or `LINE_COMMENT`, for example)

zdcthomas commented:
Alright! s/postfix_match/POSTFIX_MATCH/g is done

L3MON4D3 commented:
Perfect :ok_hand:
I'll take one last look at everything, merge #464 and then you can squash+rebase this into 1 or 2 commits on top of that. Sounds good?
BTW really nice job here, love to see PRs this clean :)

L3MON4D3 commented on DOC.md:
```diff
@@ -378,6 +379,96 @@ s("trig", {
+Some other match strings, including the default, are available from the postfix
+module. `require("luasnip.extras.postfix).matches`:
+
+- `default`: [%w%.%_%-%"%']+$
+- 'line': `^.+$`
```
I think "line" and the default-pattern are also supposed to be inside backticks, right?

zdcthomas replied:
yup!

zdcthomas replied:
i was also thinking of taking out the `"` and `'` in the default pattern, since it wouldn't match on something like `"a string with spaces in it"`

L3MON4D3 replied:
`[%w%.%_%-%"%']` would match on `"somestring"`, wouldn't it? That's alright I think

Oh an interesting addition here would be treesitter-awareness, to make the postfix capture the node before the trigger. No pressure to implement this though, I just wanted to throw the thought out there

zdcthomas replied:
Yeah I've been playing with that idea too. I think in another pr I might give it a shot, but I'd like to see how people use/ need from the normal postfix snips first I think

L3MON4D3 commented:
Okay, apart from [this](https://github.com/L3MON4D3/LuaSnip/pull/463#discussion_r900707660), we're golden :+1:
(If you'd rather just leave it at `tbl_deep_extend`, that's alright too)

zdcthomas commented:
I was looking at it, and I think it's going to end up being cleaner with the `tbl_deep_extend` but it's 100% up to you, I can put a quick helper in utils.utils with the interface you were talking about.

L3MON4D3 commented:
Naah, it's alright as-is, `tbl_set` vs `tbl_deep_extend` doesn't make a big difference

zdcthomas commented:
Sounds good to me 👍

zdcthomas commented:
Oh should I add a mention of postfix to the readme?

zdcthomas commented:
just in the features section? 
```
# Features
- Tabstops
- Text-Transformations using Lua functions
- Conditional Expansion
- Defining nested Snippets
- Filetype-specific Snippets
- Choices
- Dynamic Snippet creation
- Regex-Trigger
- Autotriggered Snippets
- Easy Postfix Snippets
- Fast
- Parse [LSP-Style](https://microsoft.github.io/language-server-protocol/specification#snippet_syntax) Snippets either directly in lua, as a vscode package or a snipmate snippet collection.
- Expand LSP-Snippets with [nvim-compe](https://github.com/hrsh7th/nvim-compe) (or its' successor, [nvim-cmp](https://github.com/hrsh7th/nvim-cmp) (requires [cmp_luasnip](https://github.com/saadparwaiz1/cmp_luasnip)))
- Snippet history (jump back into older snippets)
- Resolve filetype at the cursor using Treesitter
```

L3MON4D3 commented:
Oh yeah, please do

zdcthomas commented:
Alright, that's added to the readme. Hope that looks good!

L3MON4D3 commented:
LGTM :ok_hand: 
squash it into just a few commits, then it's g2g

zdcthomas commented:
Alright, squashed down and gtg hopefully!

L3MON4D3 commented:
Aaalmost, 1390052 contains some changes from other commits, eg. the `pre_expand_env` and a fix for the `otf`-snippets.
It would also be pretty nice if the changes to `luasnip.txt` weren't included (but this isn't necessary). (maybe do a soft reset and filter out all changes other than `extras/postfix.lua`, `DOC.md` and that test-file)

zdcthomas commented:
Ah yeah, sorry I totally missed that, I went one commit too recent with the reset. Also, should luasnip.txt be added to the gitignore? r is it sometimes updated manually as well?

L3MON4D3 commented:
Nice, that's perfect :+1:
> Also, should luasnip.txt be added to the gitignore? or is it sometimes updated manually as well?

`.gitignore` would be the best solution for it, but since this should be usable after a single clone it's simpler to just check in git. (also no, it isn't updated manually, just by that workflow)

Okay I'm merging, thank you very much for your patience+work here :D 
This is a great feature :+1:

zdcthomas commented:
Woo! thank you for your patience and help with my first contribution! It definitely won't be my last!
